reference/src/representation/enums.md

```diff
@@ -0,0 +1,342 @@
+
+The rules for enum layout vary depending on the category.
+
+### Layout of an empty enum
+
+An **empty enum** is an enum with no variants; empty enums can never
+be instantiated and are logically equivalent to the "never type"
+`!`. `#[repr]` annotations are not accepted on empty enums. Empty
+enums are guaranteed to have the same layout as `!` (zero size and
+alignment 1).
+
+### Layout of a fieldless enum
+
+If there is no `#[repr]` attached to a fieldless enum, it is
+guaranteed to be represented as an integer of sufficient size to store
+the discriminants for all possible variants. The size is selected by
+the compiler but must be at least a `u8`.
+
+When a `#[repr(Int)]`-style annotation is attached to a fieldless enum
+(one without any data for its variants), it will cause the enum to be
+represented as a simple integer of the specified size `Int`. This must
+be sufficient to store all the required discriminant values.
+
+The `#[repr(C)]` annotation is equivalent, but it selects the same
+size as the C compiler would use for the given target for an
+equivalent C-enum declaration.
+
+Combining a `C` and `Int` representation (e.g., `#[repr(C, u8)]`) is
+not permitted on a fieldless enum.
+
+The values used for the discriminant will match up with what is
+specified (or automatically assigned) in the enum definition. For
+example, the following enum defines the discriminants for its variants
+as 22 and 23 respectively:
+
+```rust
+enum Foo { 
+  // Specificy discriminant of this variant as 22:
+  Variant22 = 22,
+  
+  // Default discriminant is one more than the previous,
+  // so 23 will be assigned.
+  Variant23
+}
+```
+
+**Unresolved question:** What about platforms where `-fshort-enums`
+are the default? Do we know/care about that?
+
+### Layout of a data-carrying enums with an explicit repr annotation
+
+This section concerns data-carrying enums **with an explicit repr
+annotation of some form**. The memory layout of such cases was
+specified in [RFC 2195][] and is therefore normative.
+
+[RFC 2195]: https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html
+
+The layout of data-carrying enums that do **not** have an explicit
+repr annotation is generally undefined, but with certain specific
+exceptions: see the next section for details.
+
+#### Explicit repr annotation without C compatibility
+
+When an enum is tagged with `#[repr(Int)]` for some integral type
+`Int` (e.g., `#[repr(u8)]`), it will be represented as a C-union of a
+series of `#[repr(C)]` structs, one per variant. Each of these structs
+begins with an integral field containing the **discriminant**, which
+specifies which variant is active. They then contain the remaining
+fields associated with that variant.
+
+**Example.** The following enum uses an `repr(u8)` annotation:
+
+```rust
+#[repr(u8)]
+enum TwoCases {
+    A(u8, u16),
+    B(u16),
+}
+```
+
+This will be laid out equivalently to the following more 
+complex Rust types:
+
+```
+union TwoCasesRepr {
```

Comment:
No `repr(C)` here? AFAIK we don't otherwise guarantee the offset to be 0.

Reply:
That text was copied from the RFC. I suppose it may be incorrect. Or maybe I missed a line.